Add advice about read-only lists

atanassov opened the pull request:
The added recommendation is about the types of static lists to be considered for API design.
Addresses https://github.com/w3ctag/design-principles/issues/50


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/292.html" title="Last updated on May 13, 2021, 12:32 AM UTC (a0b49a5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/292/e5aafa3...a0b49a5.html" title="Last updated on May 13, 2021, 12:32 AM UTC (a0b49a5)">Diff</a>

LeaVerou commented:
What happened to ReadOnlyArray? Was it folded into ObservableArray as a special case?

cynthia commented:
@plinss and I discussed this in a breakout during the F2F. We think the ObservableArray recommendation is a bit strange - the observable properties are definitely not desirable by every case where an attribute is needed, and there are cases where read-only is desirable (e.g. in an attribute) - but this PR doesn't seem to cover that.

Also, @plinss mentioned he wasn't so sure about the return being restricted to only sequences. There are cases where other array types are more desirable, but this feels a bit restrictive.

annevk commented:
@LeaVerou I'm not aware of a `ReadOnlyArray` proposal, but https://heycam.github.io/webidl/#idl-observable-array does indeed satisfy that use case (it's basically up to API designers to decide what kind of operations they permit).

@domenic was it considered to have default implementations of delete and set an indexed value so that the readonly case would be consistent across APIs? That does strike me as something that is missing, but I might also be missing something. 😊

atanassov commented:
Had another round of CR with @plinss and @cynthia and arrived with the final wordsmithing addressing all concerns for this PR. Thank you all for the time.
